src/V3LinkResolve.cpp

```diff
@@ -147,38 +147,19 @@ class LinkResolveVisitor final : public VNVisitor {
     }
 
     virtual void visit(AstSenItem* nodep) override {
-        // Remove bit selects, and bark if it's not a simple variable
-        iterateChildren(nodep);
-        if (!nodep->isClocked()) {
-            // Old V1995 sensitivity list; we'll probably mostly ignore
-            bool did = true;
-            while (did) {
-                did = false;
-                if (AstNodeSel* const selp = VN_CAST(nodep->sensp(), NodeSel)) {
-                    AstNode* const fromp = selp->fromp()->unlinkFrBack();
-                    selp->replaceWith(fromp);
-                    VL_DO_DANGLING(selp->deleteTree(), selp);
-                    did = true;
-                }
-                // NodeSel doesn't include AstSel....
-                if (AstSel* const selp = VN_CAST(nodep->sensp(), Sel)) {
-                    AstNode* const fromp = selp->fromp()->unlinkFrBack();
-                    selp->replaceWith(fromp);
-                    VL_DO_DANGLING(selp->deleteTree(), selp);
-                    did = true;
-                }
-                if (AstNodePreSel* const selp = VN_CAST(nodep->sensp(), NodePreSel)) {
-                    AstNode* const fromp = selp->fromp()->unlinkFrBack();
-                    selp->replaceWith(fromp);
-                    VL_DO_DANGLING(selp->deleteTree(), selp);
-                    did = true;
-                }
-            }
-        }
-        if (nodep->isIllegal()) {
-            if (debug()) nodep->dumpTree(cout, "-tree: ");
-            nodep->v3warn(E_UNSUPPORTED, "Unsupported: Complex statement in sensitivity list");
+        UASSERT_OBJ(nodep->isClocked(), nodep, "Invalid edge");
+        // Split concatenations into multiple senitems (only if the edge is ET_CHANGED)
+        if (nodep->edgeType() == VEdgeType::ET_CHANGED
+            && (VN_IS(nodep->sensp(), Concat) || VN_IS(nodep->sensp(), Replicate))) {
```

Comment:
What about concat in concat, replicate in concat, etc?  Some recursion and tests needed?

Reply:
From what I understand, concat in Verilator is a binary op, and concat of more than 2 elements is a concat in concat.

As for recursion, this `visit` flattens a concat/replicate tree into a flat sentree. Each element is put in a new senitem which either replaces the current one, or is added after it. The new senitems are visited and the process repeats, until there are no more concats/replicates.

Reply:
My point was to add to tests e.g. "{{siga, sigb}, {sigc, 2{sigd}}" and see if that is handled.  From my reading of the code you are using an "if" so only a single concat will work and the rest won't be handled. By recursion I meant you may need to recursively call a function instead of the "if" to blow out the concats/replicates as there's a lot of possibilities that are ugly to handle with some while loops but might be easier with recursion.  Or maybe a while loop works fine, you'll know when you try it.   Of course maybe I'm misreading and once you add the test it'll just pass, then you are done for this comment.

Reply:
I understood what you meant, but recursion or a loop isn't needed here. You're right about the test, though, I added one.

I thought this was covered by some existing test as Verilator does support this sort of thing on `master`: https://github.com/verilator/verilator/blob/db5fdfb0ee871e4cf8a25b8890207b507d5fa7c9/src/verilog.y#L3030

Reply:
It's just optimization, and I put it there to keep the behavior as close as possible to how it worked before this patch. I'll find a better place for it. And I'll try the `posedge` optimization, though I can't promise I'll find the time to implement it.

Reply:
For the edge thing, I conjecture you could replace the expr with AstSel(expr, 0, 1), and the rest of the constant folding should take care of it.

Reply:
> If you just want to reduce the number of unique expressions (and hence the size of the trigger vec), 

Note to self: And of course I'm speaking nonsense as the triggervec size depends on the number of unique SenTrees not SenItems, still you can have fewer "prev" values stored by splitting them up, if they happen to be reused.

Reply:
Initially I tried your `AstSel` suggestion, and put concat optimization in `V3Const`. It worked, but it resulted in a `WIDTHCONCAT` warning in `t_alw_combdly`. So the concats have to be simplified before or in `V3Width` (that's where I put it for now).